### PR TITLE
Adds connect and read timeouts to discovery-gce

### LINKED ISSUE #24313: GceDiscoverTests fails with thread leak
build URL:
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.3+multijob-unix-compatibility/os=opensuse/144/consoleFull

Reproduce command:
```
gradle :plugins:discovery-gce:test -Dtests.seed=AD2321B05BCFD7A9 -Dtests.class=org.elasticsearch.discovery.gce.GceDiscoverTests -Dtests.security.manager=true -Dtests.locale=en-US -Dtests.timezone=Etc/UTC
```

stack trace:
```
  2> REPRODUCE WITH: gradle :plugins:discovery-gce:test -Dtests.seed=AD2321B05BCFD7A9 -Dtests.class=org.elasticsearch.discovery.gce.GceDiscoverTests -Dtests.security.manager=true -Dtests.locale=en-US -Dtests.timezone=Etc/UTC
  2> REPRODUCE WITH: gradle :plugins:discovery-gce:test -Dtests.seed=AD2321B05BCFD7A9 -Dtests.class=org.elasticsearch.discovery.gce.GceDiscoverTests -Dtests.security.manager=true -Dtests.locale=en-US -Dtests.timezone=Etc/UTC
  2> NOTE: test params are: codec=Asserting(Lucene62): {}, docValues:{}, maxPointsInLeafNode=1668, maxMBSortInHeap=6.040847236532724, sim=RandomSimilarity(queryNorm=false,coord=crazy): {}, locale=en-MT, timezone=America/Cordoba
  2> NOTE: Linux 3.16.7-53-default amd64/Oracle Corporation 1.8.0_111 (64-bit)/cpus=4,threads=2,free=369534272,total=514850816
  2> NOTE: All tests run in this JVM: [GceDiscoverTests]
ERROR   0.00s J0 | GceDiscoverTests (suite) <<< FAILURES!
   > Throwable #1: com.carrotsearch.randomizedtesting.ThreadLeakError: 1 thread leaked from SUITE scope at org.elasticsearch.discovery.gce.GceDiscoverTests: 
   >    1) Thread[id=41, name=elasticsearch[node_s0][generic][T#2], state=RUNNABLE, group=TGRP-GceDiscoverTests]
   >         at java.net.SocketInputStream.socketRead0(Native Method)
   >         at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
   >         at java.net.SocketInputStream.read(SocketInputStream.java:170)
   >         at java.net.SocketInputStream.read(SocketInputStream.java:141)
   >         at sun.security.ssl.InputRecord.readFully(InputRecord.java:465)
   >         at sun.security.ssl.InputRecord.read(InputRecord.java:503)
   >         at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:973)
   >         at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1375)
   >         at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1403)
   >         at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1387)
   >         at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:559)
   >         at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:185)
   >         at sun.net.www.protocol.https.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java:153)
   >         at com.google.api.client.http.javanet.NetHttpRequest.execute(NetHttpRequest.java:93)
   >         at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:972)
   >         at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:419)
   >         at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352)
   >         at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469)
   >         at org.elasticsearch.cloud.gce.GceInstancesServiceImpl$1.run(GceInstancesServiceImpl.java:79)
   >         at org.elasticsearch.cloud.gce.GceInstancesServiceImpl$1.run(GceInstancesServiceImpl.java:75)
   >         at java.security.AccessController.doPrivileged(Native Method)
   >         at org.elasticsearch.cloud.gce.GceInstancesServiceImpl.lambda$instances$1(GceInstancesServiceImpl.java:75)
   >         at org.elasticsearch.cloud.gce.GceInstancesServiceImpl$$Lambda$1288/1577691077.apply(Unknown Source)
   >         at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
   >         at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374)
   >         at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
   >         at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
   >         at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
   >         at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
   >         at java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:474)
   >         at org.elasticsearch.cloud.gce.GceInstancesServiceImpl.instances(GceInstancesServiceImpl.java:91)
   >         at org.elasticsearch.discovery.gce.GceUnicastHostsProvider.buildDynamicNodes(GceUnicastHostsProvider.java:134)
   >         at org.elasticsearch.discovery.zen.UnicastZenPing.ping(UnicastZenPing.java:302)
   >         at org.elasticsearch.discovery.zen.UnicastZenPing.ping(UnicastZenPing.java:279)
   >         at org.elasticsearch.discovery.zen.ZenDiscovery.pingAndWait(ZenDiscovery.java:1007)
   >         at org.elasticsearch.discovery.zen.ZenDiscovery.findMaster(ZenDiscovery.java:874)
   >         at org.elasticsearch.discovery.zen.ZenDiscovery.innerJoinCluster(ZenDiscovery.java:383)
   >         at org.elasticsearch.discovery.zen.ZenDiscovery.access$3800(ZenDiscovery.java:81)
   >         at org.elasticsearch.discovery.zen.ZenDiscovery$JoinThreadControl$1.run(ZenDiscovery.java:1190)
   >         at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:569)
   >         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
   >         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
   >         at java.lang.Thread.run(Thread.java:745)
   > 	at __randomizedtesting.SeedInfo.seed([AD2321B05BCFD7A9]:0)Throwable #2: com.carrotsearch.randomizedtesting.ThreadLeakError: There are still zombie threads that couldn't be terminated:
   >    1) Thread[id=41, name=elasticsearch[node_s0][generic][T#2], state=RUNNABLE, group=TGRP-GceDiscoverTests]
   >         at java.net.SocketInputStream.socketRead0(Native Method)
   >         at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
   >         at java.net.SocketInputStream.read(SocketInputStream.java:170)
   >         at java.net.SocketInputStream.read(SocketInputStream.java:141)
   >         at sun.security.ssl.InputRecord.readFully(InputRecord.java:465)
   >         at sun.security.ssl.InputRecord.read(InputRecord.java:503)
   >         at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:973)
   >         at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1375)
   >         at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1403)
   >         at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1387)
   >         at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:559)
   >         at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:185)
   >         at sun.net.www.protocol.https.HttpsURLConnectionImpl.connect(HttpsURLConnectionImpl.java:153)
   >         at com.google.api.client.http.javanet.NetHttpRequest.execute(NetHttpRequest.java:93)
   >         at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:972)
   >         at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:419)
   >         at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352)
   >         at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469)
   >         at org.elasticsearch.cloud.gce.GceInstancesServiceImpl$1.run(GceInstancesServiceImpl.java:79)
   >         at org.elasticsearch.cloud.gce.GceInstancesServiceImpl$1.run(GceInstancesServiceImpl.java:75)
   >         at java.security.AccessController.doPrivileged(Native Method)
   >         at org.elasticsearch.cloud.gce.GceInstancesServiceImpl.lambda$instances$1(GceInstancesServiceImpl.java:75)
   >         at org.elasticsearch.cloud.gce.GceInstancesServiceImpl$$Lambda$1288/1577691077.apply(Unknown Source)
   >         at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
   >         at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374)
   >         at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
   >         at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
   >         at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
   >         at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
   >         at java.util.stream.ReferencePipeline.reduce(ReferencePipeline.java:474)
   >         at org.elasticsearch.cloud.gce.GceInstancesServiceImpl.instances(GceInstancesServiceImpl.java:91)
   >         at org.elasticsearch.discovery.gce.GceUnicastHostsProvider.buildDynamicNodes(GceUnicastHostsProvider.java:134)
   >         at org.elasticsearch.discovery.zen.UnicastZenPing.ping(UnicastZenPing.java:302)
   >         at org.elasticsearch.discovery.zen.UnicastZenPing.ping(UnicastZenPing.java:279)
   >         at org.elasticsearch.discovery.zen.ZenDiscovery.pingAndWait(ZenDiscovery.java:1007)
   >         at org.elasticsearch.discovery.zen.ZenDiscovery.findMaster(ZenDiscovery.java:874)
   >         at org.elasticsearch.discovery.zen.ZenDiscovery.innerJoinCluster(ZenDiscovery.java:383)
   >         at org.elasticsearch.discovery.zen.ZenDiscovery.access$3800(ZenDiscovery.java:81)
   >         at org.elasticsearch.discovery.zen.ZenDiscovery$JoinThreadControl$1.run(ZenDiscovery.java:1190)
   >         at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:569)
   >         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
   >         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
   >         at java.lang.Thread.run(Thread.java:745)
   > 	at __randomizedtesting.SeedInfo.seed([AD2321B05BCFD7A9]:0)
```

### PR DESCRIPTION
**Pull request by tlrx**
This commit adds two new settings to the `discovery-gce` plugin
(`cloud.gce.connect_timeout` and `cloud.gce.read_timeout`) that
allow to configure the socket's connection timeout and read timeout
when accessing to Google's Compute Engine APIs.

Similar settings already exist in discovery-ec2, repository-gcs, 
repository-s3 and repository-azure plugins.

This commit refactor the RetryHttpRequestInitializerWrapper class, 
removes an unused GceModule class and use the new settings to 
fix #24313 that fails every two months or so. The GceDiscoverTests 
is also changed to not rely on transports ports file anymore but directly 
register nodes in the test instead.

### CODE REVIEW

File: docs/plugins/discovery-gce.asciidoc [hunk 1]
```diff
@@ -34,6 +34,18 @@ The following gce settings (prefixed with `cloud.gce`) are supported:
      https://developers.google.com/compute/docs/zones#available[GCE supported zones].
      See also <<discovery-gce-usage-zones>>.
 
+ `connect_timeout`::
+
+     Sets the timeout to establish a connection. A value of `0` corresponds to an infinite timeout, while
+     a value of `-1` corresponds to the default timeout provided by the Google Compute Engine Java Library (20 seconds).
+     Defaults to `-1`.
```
Comment (bleskes) [inline docs/plugins/discovery-gce.asciidoc]: why can't we use `null` for unassigned/default and -1 for infinity?

**Review by bleskes — CHANGES_REQUESTED** [review]
Looks like a good cleanup. I left a question about the settings

File: docs/plugins/discovery-gce.asciidoc (hunk 1 above)
Comment (tlrx) [reply docs/plugins/discovery-gce.asciidoc]: Right, that would be aligned with the max_wait setting. I'll change that.

**Comment by tlrx** [comment]
@bleskes I updated the settings.

File: plugins/discovery-gce/src/main/java/org/elasticsearch/discovery/gce/GceHttpRequestInitializer.java [hunk 2]
```diff
@@ -90,16 +90,23 @@ public void initialize(HttpRequest httpRequest) {
...
-
-        final long connectTimeoutMillis = connectTimeout.getMillis();
-        if (connectTimeoutMillis > noTimeout) {
-            httpRequest.setConnectTimeout((int) connectTimeoutMillis);
+        final long infiniteTimeout = TimeValue.MINUS_ONE.getMillis();
+        if (connectTimeout != null) {
+            final long connectTimeoutMillis = connectTimeout.getMillis();
+            if (connectTimeoutMillis > 0) {
+                httpRequest.setConnectTimeout((int) connectTimeoutMillis);
+            } else if (connectTimeoutMillis == infiniteTimeout) {
```
Comment (bleskes) [inline plugins/discovery-gce/src/main/java/org/elasticsearch/discovery/gce/GceHttpRequestInitializer.java]: this smells like it should be a setting validation thing. Testing for this so deep, on every request without throwing exceptions feels wrong.

**Review by bleskes — CHANGES_REQUESTED** [review]
Thanks Tanguy. I left one more comment

File: plugins/discovery-gce/src/main/java/org/elasticsearch/discovery/gce/GceHttpRequestInitializer.java (hunk 2 above)
Comment (tlrx) [reply plugins/discovery-gce/src/main/java/org/elasticsearch/discovery/gce/GceHttpRequestInitializer.java]: Raaah yes you're right, I should have seen that. Thanks.

**Comment by tlrx** [comment]
@elasticmachine retest this please

**Comment by tlrx** [comment]
@elasticmachine retest this please

**Comment by tlrx** [comment]
@bleskes Can you please have another look when you have time? Thanks

**Comment by ywelsch** [comment]
@tlrx is this still something we want to get in?

**Comment by ywelsch** [comment]
@tlrx while we would still be interested in this PR, it will need to be brought up-to-date with latest master. I'm closing this one out due to inactivity, but feel free to reopen.

**Pull request closed without merging**